File: app/src/main/java/jp/co/pannacotta/lunch_app/ResultActivity.java
package jp.co.pannacotta.lunch_app;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.text.method.MovementMethod;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.io.IOException;
import java.util.Random;

import androidx.appcompat.app.AppCompatActivity;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.preference.PreferenceManager;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;

import jp.co.pannacotta.lunch_app.model.Gourmet;
import jp.co.pannacotta.lunch_app.model.Shop;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ResultActivity extends AppCompatActivity {
    private MediaPlayer mediaPlayer;
    public static final String API_URL = "http://webservice.recruit.co.jp";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_result);

        audioPlay();

        TextView CregitTextView = findViewById(R.id.CregitTextView);
        String htmlcredit = "【画像提供：ホットペッパー グルメ】<br>Powered by <a href=\"http://webservice.recruit.co.jp/\">ホットペッパー Webサービス</a>";

        CharSequence cregitChar = Html.fromHtml(htmlcredit);

        CregitTextView.setText(cregitChar);

        MovementMethod mMethod = LinkMovementMethod.getInstance();
        CregitTextView.setMovementMethod(mMethod);

        Button button = findViewById(R.id.next_shop_botton);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent();
                intent.setClass(ResultActivity.this, AngryActivity.class);
                startActivity(intent);
                mediaPlayer.stop();
            }
        });

        Button web_button = findViewById(R.id.shop_web_botton);
        web_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mediaPlayer.stop();
            }
        });

        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.addInterceptor(new Interceptor() {
            @Override
            public okhttp3.Response intercept(Chain chain) throws IOException {
                Request original = chain.request();

                //header設定
                Request request = original.newBuilder()
                        .header("Accept", "application/json")
                        .method(original.method(), original.body())
                        .build();

                okhttp3.Response response = chain.proceed(request);
                return response;
            }
        });

        //ログ出力設定
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BASIC);
        httpClient.addInterceptor(logging);

        //クライアント生成
        OkHttpClient client = httpClient.build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();

        Hotpepperapi service = retrofit.create(Hotpepperapi.class);
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        String lat = pref.getString("lat", "");
        String lng = pref.getString("lng", "");
        Call<Gourmet> call = service.webservice("ae27b52bed304677", lat, lng, "3", "1", "100", "json");
        call.enqueue(new Callback<Gourmet>() {
            @Override
            public void onResponse(Call<Gourmet> call, Response<Gourmet> response) {
                Log.d("ろぐ", response.toString());
                int shop_count = response.body().results.shop.size();
                if (shop_count > 0) {
                    Random random = new Random();
                    int todayShopnum = random.nextInt(shop_count - 1);
                    final Shop todayShop = response.body().results.shop.get(todayShopnum);

                    String todayShopName = todayShop.name;
                    ImageView shop_photo = findViewById(R.id.shop_photo);
                    String todayShopPhotoUrl = todayShop.photo.pc.l;
                    Glide.with(ResultActivity.this).load(todayShopPhotoUrl).apply(new RequestOptions().override(700, 1000)).into(shop_photo);

                    String todayShopCatchCopy = todayShop.catchCopy;

                    TextView shop_name = findViewById(R.id.shop_name);

                    TextView shop_catch_copy = findViewById(R.id.shop_catch_copy);

                    shop_name.setText(todayShopName);


                    shop_catch_copy.setText(todayShopCatchCopy);

                    Button button = findViewById(R.id.shop_web_botton);
                    button.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            String todayShopUrl = todayShop.urls.pc;
                            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
                            CustomTabsIntent customTabsIntent = builder.build();
                            customTabsIntent.launchUrl(ResultActivity.this, Uri.parse(todayShopUrl));
                        }
                    });
                } else {
                    goErrorActivity();
                }
            }

            @Override
            public void onFailure(Call<Gourmet> call, Throwable t) {
                goErrorActivity();
            }
        });

    }

    private void goErrorActivity() {
        Intent intent = new Intent();
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.setClass(ResultActivity.this, ErrorActivity.class);
        startActivity(intent);
    }
    private boolean audioSetup() {
        boolean fileCheck = false;

        mediaPlayer = new MediaPlayer();

        String filePath = "lunch.wav";

        try (AssetFileDescriptor afdescripter = getAssets().openFd(filePath);) {
            // MediaPlayerに読み込んだ音楽ファイルを指定
            mediaPlayer.setDataSource(afdescripter.getFileDescriptor(),
                    afdescripter.getStartOffset(),
                    afdescripter.getLength());
            // 音量調整を端末のボタンに任せる
            setVolumeControlStream(AudioManager.STREAM_MUSIC);
            mediaPlayer.prepare();
            mediaPlayer.setLooping(true);
            fileCheck = true;
        } catch (IOException e1) {
            e1.printStackTrace();
        }
        return fileCheck;
    }

    private void audioPlay() {

        if (mediaPlayer == null) {
            if (audioSetup()) {

            } else {
                return;
            }
        } else {
            mediaPlayer.stop();
            mediaPlayer.reset();
            mediaPlayer.release();
        }
        // 再生する
        mediaPlayer.start();
    }

    @Override
    public void onPause() {
        super.onPause();
        mediaPlayer.pause();
    }

    @Override
    public void onRestart() {
        super.onRestart();
        mediaPlayer.start();
    }
}

File: app/src/main/java/jp/co/pannacotta/lunch_app/ResultActivity.kt
package jp.co.pannacotta.lunch_app

import android.content.Intent
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.SoundPool
import android.net.Uri
import android.os.Bundle
import android.text.Html
import android.text.method.LinkMovementMethod
import android.util.Log
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.browser.customtabs.CustomTabsIntent
import androidx.preference.PreferenceManager
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import jp.co.pannacotta.lunch_app.model.Gourmet
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.util.*

class ResultActivity : AppCompatActivity() {
    private var mediaPlayer: MediaPlayer? = null
    private var htmlcreditText: String? = null
    private var webservice_key: String? = null
    private var webservice_range: String? = null
    private var webservice_lunch: String? = null
    private var webservice_count: String? = null
    private var webservice_format: String? = null
    private var soundPool: SoundPool? = null
    private var voice_koko = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_result)
        htmlcreditText = getString(R.string.html_credit)
        webservice_key = getString(R.string.webservice_key)
        webservice_range = getString(R.string.webservice_range)
        webservice_lunch = getString(R.string.webservice_lunch)
        webservice_count = getString(R.string.webservice_count)
        webservice_format = getString(R.string.webservice_format)
        val audioAttributes = AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_GAME).setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).build()
        soundPool = SoundPool.Builder().setAudioAttributes(audioAttributes).setMaxStreams(2).build()
        voice_koko = soundPool!!.load(this,R.raw.koko,1)

        soundPool!!.setOnLoadCompleteListener { soundPool, sampleId, status ->
            if (0 == status) {
                when  (sampleId) {
                    voice_koko -> soundPool!!.play(voice_koko, 0.3f, 0.3f, 0, 0, 1.0f)
                }
            }
        }
        audioPlay()
        val CregitTextView = findViewById<TextView>(R.id.CregitTextView)
        val htmlcredit = htmlcreditText
        val cregitChar: CharSequence = Html.fromHtml(htmlcredit)
        CregitTextView.text = cregitChar
        val mMethod = LinkMovementMethod.getInstance()
        CregitTextView.movementMethod = mMethod
        val button = findViewById<Button>(R.id.next_shop_button)
        button.setOnClickListener {
            val intent = Intent()
            intent.setClass(this@ResultActivity, AngryActivity::class.java)
            startActivity(intent)
            mediaPlayer!!.stop()
        }
        val web_button = findViewById<Button>(R.id.shop_web_button)
        web_button.setOnClickListener { mediaPlayer!!.stop() }
        val httpClient = OkHttpClient.Builder()
        httpClient.addInterceptor { chain ->
            val original = chain.request()

            //header設定
            val request = original.newBuilder()
                    .header("Accept", "application/json")
                    .method(original.method(), original.body())
                    .build()
            chain.proceed(request)
        }

        //ログ出力設定
        val logging = HttpLoggingInterceptor()
        logging.level = HttpLoggingInterceptor.Level.BASIC
        httpClient.addInterceptor(logging)

        //クライアント生成
        val client = httpClient.build()
        val retrofit = Retrofit.Builder()
                .baseUrl(API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build()
        val service = retrofit.create(Hotpepperapi::class.java)
        val pref = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        val lat = pref.getString("lat", "")
        val lng = pref.getString("lng", "")
        val call = service.webservice(webservice_key, lat, lng, webservice_range, webservice_lunch, webservice_count, webservice_format)
        call!!.enqueue(object : Callback<Gourmet?> {
            override fun onResponse(call: Call<Gourmet?>, response: retrofit2.Response<Gourmet?>) {
                Log.d("ろぐ", response.toString())
                val shop_count = response.body()!!.results!!.shop!!.size
                if (shop_count > 0) {
                    val random = Random()
                    val todayShopnum = random.nextInt(shop_count - 1)
                    val todayShop = response.body()!!.results!!.shop!![todayShopnum]
                    val todayShopName = todayShop.name
                    val shop_photo = findViewById<ImageView>(R.id.shop_photo)
                    val todayShopPhotoUrl = todayShop.photo!!.pc!!.l
                    Glide.with(this@ResultActivity).load(todayShopPhotoUrl).apply(RequestOptions().override(1000, 1000)).into(shop_photo)
                    val todayShopCatchCopy = todayShop.catchCopy
                    val shop_name = findViewById<TextView>(R.id.shop_name)
                    val shop_catch_copy = findViewById<TextView>(R.id.shop_catch_copy)
                    shop_name.text = todayShopName
                    shop_catch_copy.text = todayShopCatchCopy
                    val button = findViewById<Button>(R.id.shop_web_button)
                    button.setOnClickListener {
                        val todayShopUrl = todayShop.urls!!.pc
                        val builder = CustomTabsIntent.Builder()
                        val customTabsIntent = builder.build()
                        customTabsIntent.launchUrl(this@ResultActivity, Uri.parse(todayShopUrl))
                    }
                } else {
                    goErrorActivity()
                }
            }

            override fun onFailure(call: Call<Gourmet?>, t: Throwable) {
                goErrorActivity()
            }
        })
    }

    private fun goErrorActivity() {
        val intent = Intent()
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        intent.setClass(this@ResultActivity, ErrorActivity::class.java)
        startActivity(intent)
    }

    private fun audioSetup(): Boolean {
        var fileCheck = false
        mediaPlayer = MediaPlayer()
        val filePath = "lunch.wav"
        try {
            assets.openFd(filePath).use { afdescripter ->
                // MediaPlayerに読み込んだ音楽ファイルを指定
                mediaPlayer!!.setDataSource(afdescripter.fileDescriptor,
                        afdescripter.startOffset,
                        afdescripter.length)
                // 音量調整を端末のボタンに任せる
                volumeControlStream = AudioManager.STREAM_MUSIC
                mediaPlayer!!.prepare()
                mediaPlayer!!.isLooping = true
                fileCheck = true
            }
        } catch (e1: IOException) {
            e1.printStackTrace()
        }
        return fileCheck
    }

    private fun audioPlay() {
        if (mediaPlayer == null) {
            if (audioSetup()) {
            } else {
                return
            }
        } else {
            mediaPlayer!!.stop()
            mediaPlayer!!.reset()
            mediaPlayer!!.release()
        }
        // 再生する
        mediaPlayer!!.start()
    }

    fun audioStop() {
        // 再生終了
        mediaPlayer?.stop()
        // リセット
        mediaPlayer?.reset()
        // リソースの解放
        mediaPlayer?.release()
        mediaPlayer = null
    }

    public override fun onPause() {
        super.onPause()
        audioStop()
        soundPool?.stop(1)
    }

    public override fun onRestart() {
        super.onRestart()
        audioPlay()
    }

    companion object {
        const val API_URL = "http://webservice.recruit.co.jp"
    }
}